### PR TITLE
New concept article about the key-param challenge

### PR DESCRIPTION
**Pull request by andikrueger**
#### Pull Request (PR) description

This PR will add a new concept page to outline the challenges with key parameters in DSC and M365.

### CODE REVIEW

File: docs/docs/concepts/key-parameters.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+## The current implementation of key parameters in Microsoft365DSC
+
+With the April 2023 release of Microsoft365DSC we have many breaking changes that help to better identify the resources in the configuration. There were several changes to key parameters. This implementation  a new way of handling key parameters. This new implementation is based on the following principles:
+
+1. The key parameters are the parameters that are used to identify a resource.
+2. The immutable identifiers are the unique identifiers that are created during the creation of the resource. Fom some resources, these are now exposed through M365DSC. This offers two use cases:
+   1. During the initial creation of the resource the display name would be used to identify the resource. This is the same as the current implementation.
+   2. For updates of the resource, the immutable identifier would be used to identify the resource. This is a new implementation.
+
+The new implementation is kind of a heuristic approach to enable exports and imports of configurations. Besides that it is a good practice to better identify the resources in the configuration.
```
Comment (ykuijs) [inline docs/docs/concepts/key-parameters.md:85]: Maybe we can add a "More information" section with a link to this article:
https://learn.microsoft.com/en-us/powershell/dsc/resources/authoringresourcemof?view=dsc-1.1#the-contents-of-the-mof-file

> The type qualifier, [Key], on a property indicates that this property will uniquely identify the resource instance. At least one [Key] property is required.

**Review by ykuijs — COMMENTED** [review]
Few comments

File: docs/docs/concepts/key-parameters.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        AADUser User2 {
+            UserPrincipalName = "User1@tenant.onmicrosoft.com"
+            DisplayName       = "User2"
+        }
+    }
+}
+
+```
+
+There are two resources with the same key combination. User1 and User2 do have the same UserPrincipalName and would configure different DisplayNames. This is not allowed and will not compile. The same is true for other resources. For example, if you would have two AADGroups with the same UserPrincipalName, the configuration would not compile.
```
Comment (ykuijs) [inline docs/docs/concepts/key-parameters.md:57]: In the above examples we only have one Key parameter. Shouldn't we add the comment that multiple parameters can be marked key and that the combination must be unique within a configuration?

File: docs/docs/concepts/key-parameters.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+Changing the key parameters of those resource with duplicate keys is not an option as we would create a different configuration. In most cases the root cause is the fact, that we can have objects in Azure and Microsoft 365 with the same display name, but different immutable identifiers.
+
+We can differentiate between two types of resources:
+
+1. Those which have a unique immutable identifier and can be identified by the key parameters. Like the AADUser resource, which has a unique immutable identifier and can be identified by the UserPrincipalName.
+2. Those which have a unique immutable identifier, but can not be identified by the key parameters. Like the AADApplication resource, which has a unique immutable identifier, but can not be uniquely identified by the DisplayName.
```
[the remark this answers is not in the document]
Comment (andikrueger) [reply docs/docs/concepts/key-parameters.md:72]: done.

File: docs/docs/concepts/key-parameters.md (hunk 2 above)
Comment (andikrueger) [reply docs/docs/concepts/key-parameters.md:57]: done

File: docs/docs/concepts/key-parameters.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+For 1. we can change the key parameters and the configuration will compile. We can rely on the immutable identifier to identify the resource. For 2. we can not rely on the immutable identifier to identify the resource. This is a big challenge for DSC, as we need to have a unique identifier prior to the creation of the objects in the tenant.
+
+## The current implementation of key parameters in Microsoft365DSC
+
+With the April 2023 release of Microsoft365DSC we have many breaking changes that help to better identify the resources in the configuration. There were several changes to key parameters. This implementation  a new way of handling key parameters. This new implementation is based on the following principles:
```
[the remark this answers is not in the document]
Comment (andikrueger) [reply docs/docs/concepts/key-parameters.md:78]: done

File: docs/docs/concepts/key-parameters.md (hunk 2 above)
Comment (andikrueger) [reply docs/docs/concepts/key-parameters.md:57]: done

File: docs/docs/concepts/key-parameters.md (hunk 1 above)
Comment (andikrueger) [reply docs/docs/concepts/key-parameters.md:85]: done

File: docs/docs/concepts/key-parameters.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+For 1. we can change the key parameters and the configuration will compile. We can rely on the immutable identifier to identify the resource. For 2. we can not rely on the immutable identifier to identify the resource. This is a big challenge for DSC, as we need to have a unique identifier prior to the creation of the objects in the tenant.
+
+## The current implementation of key parameters in Microsoft365DSC
+
+With the [April 2023 release](../blog/april-2023-major-release/index.html) of Microsoft365DSC we have many breaking changes that help to better identify the resources in the configuration. There were several changes to key parameters. This implementation  a new way of handling key parameters. This new implementation is based on the following principles:
+
+1. The key parameters are the parameters that are used to identify a resource.
+   There is [more information](https://docs.microsoft.com/en-us/powershell/dsc/authoringresourcekey) available on how to define key parameters in DSC resources:
+   >The type qualifier, [Key], on a property indicates that this property will uniquely identify the resource instance. At least one [Key] property is required.
```
Comment (ykuijs) [inline docs/docs/concepts/key-parameters.md:86]: Does using a quote in a list render correctly?

Comment (andikrueger) [reply docs/docs/concepts/key-parameters.md:86]: It should if the indention is correct. 🤞

**Pull request merged**